paasta_tools/firewall_update.py

```diff
@@ -0,0 +1,114 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import argparse
+import logging
+import os.path
+import time
+from collections import defaultdict
+
+from inotify.adapters import Inotify
+from inotify.constants import IN_MODIFY
+from inotify.constants import IN_MOVED_TO
+
+from paasta_tools.chronos_tools import chronos_services_running_here
+from paasta_tools.chronos_tools import load_chronos_job_config
+from paasta_tools.marathon_tools import load_marathon_service_config
+from paasta_tools.marathon_tools import marathon_services_running_here
+from paasta_tools.utils import DEFAULT_SOA_DIR
+from paasta_tools.utils import load_system_paasta_config
+
+log = logging.getLogger(__name__)
+
+DEFAULT_UPDATE_SECS = 5
+DEFAULT_SYNAPSE_SERVICE_DIR = b'/var/run/synapse/services'
+
+
+def parse_args(argv):
+    parser = argparse.ArgumentParser(description='Monitor synapse changes and update service firewall rules')
+    parser.add_argument('--synapse-service-dir', dest="synapse_service_dir",
+                        default=DEFAULT_SYNAPSE_SERVICE_DIR,
+                        help="Path to synapse service dir (default %(default)s)")
+    parser.add_argument('-d', '--soa-dir', dest="soa_dir", metavar="soa_dir",
+                        default=DEFAULT_SOA_DIR,
+                        help="define a different soa config directory (default %(default)s)")
+    parser.add_argument('-u', '--update-secs', dest="update_secs",
+                        default=DEFAULT_UPDATE_SECS, type=int,
+                        help="Poll for new containers every N secs (default %(default)s)")
+    parser.add_argument('-v', '--verbose', dest='verbose', action='store_true')
+
+    args = parser.parse_args(argv)
+    return args
+
+
+def setup_logging(verbose):
+    level = logging.DEBUG if verbose else logging.WARNING
+    logging.basicConfig(level=level)
+
+
+def run(args):
+    # Main loop waiting on inotify file events
+    inotify = Inotify(block_duration_s=1)  # event_gen blocks for 1 second
+    inotify.add_watch(args.synapse_service_dir.encode(), IN_MOVED_TO | IN_MODIFY)
```

Comment:
I think you are going to need to watch service.yaml too.

When it changes, you will need to *also* update firewall rules, but also you need to break the yaml cache for that (`service_configuration_lib._yaml_cache = {}`) otherwise it will read in a cached value.